except_star.md

```diff
@@ -234,20 +284,177 @@ except *ValueError as errors:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  print(f'got some TypeErrors: {len(e)}')
+except *Exception:
+  pass
+
+# would print:
+#
+#  got some TypeErrors: 2
+```
+
+Iteration over an `ExceptionGroup` that has nested `ExceptionGroup` objects
+in it effectively flattens the entire tree. E.g.
+
+```python
+print(
+  list(
+    ExceptionGroup(
+      ValueError('a'),
+      TypeError('b'),
+      ExceptionGroup(
+        TypeError('c'),
+        KeyError('d')
+      )
+    )
+  )
+)
+
+# would output:
+#
+#   [ValueError('a'), TypeError('b'), TypeError('c'), KeyError('d')]
+```
+
+### "raise e" vs "raise"
+
+The difference between bare `raise` and a more specific `raise e` is more
+significant for exception groups than it is for regular exceptions. Consider
+the following two examples that illustrate it.
+
+Bare `raise` preserves the exception group internal structure:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  raise
+
+# would crash with:
+#
+#  ExceptionGroup(
+#    ValueError('a'),
+#    TypeError('b'),
+#    ExceptionGroup(
+#      TypeError('c'),
+#      KeyError('d')
+#    )
+#  )
+```
+
+Whereas `raise e` would flatten the captured subset:
+
+```python
+try:
+  raise ExceptionGroup(
+    ValueError('a'),
+    TypeError('b'),
+    ExceptionGroup(
+      TypeError('c'),
+      KeyError('d')
+    )
+  )
+except *TypeError as e:
+  raise e
+
+# would crash with:
+#
+#  ExceptionGroup(
+#    ValueError('a'),
+#    ExceptionGroup(
+#      TypeError('b'),
+#      TypeError('c'),
+#    ),
+#    ExceptionGroup(
+#      KeyError('d')
+#    )
+#  )
 ```
 
-### Types of errors
+### "continue" and "break" in "except*"
+
+Both `continue` and `break` are disallowed in `except*` clauses, causing
+a `SyntaxError`.
+
+Due to the fact that `try..except*` block allows multiple `except*` clauses
+to run while handling one `ExceptionGroup` with multiple different exceptions
+in it, allowing one innocent `break` or `continue` in one `except*` to
+effectively silence the entire group feels very error prone.
+
+### "return" in "except*"
+
+A `return` in a regular `except` or `finally` clause means
+"suppress the exception". For example, both of the below functions would
+silence their `ZeroDivisionError`s:
+
+```python
+def foo():
+   try:
+      1 / 0
+   finally:
+      print('the sound of')
+      return
+
+def bar():
+   try:
+      1 / 0
+   except ZeroDivisionError:
+     pass
+   finally:
+      print('silence')
+      return
+
+foo()
+bar()
+
+# would print:
+#
+#   the sound of
+#   silence
+```
+
+We propose to replicate this behavior in the `except*` syntax as it is useful
+as an escape hatch when it's clear that all exceptions can be silenced.
```

Comment:
Unless you *only* allow `return` without expression that's not enough -- what would happen here?
```
try:
    raise ExceptionGroup(A(), B())
except *A:
    return 1
except *B:
    return 2
```

Reply:
Right, great catch. I think we should only allow bare return. I'll add a clarification.

(FWIW I never use this feature myself)

Reply:
> what would happen here?

I think it would return `1`, because `A()` would be matched first. But when running async code this is too fragile and would only allow users to write unpredictable code. IMO we should disable non-None returns.

Reply:
There’s room to disagree here, but I would prefer not allowing any return statements over only allowing bare return. I worry that allowing bare return with the proposed semantics is closing the door for assigning *any* semantics to return with expression and for break/continue.

Reply:
> but I would prefer not allowing any return statements over only allowing bare return

TBH I'd prefer that too. As I said earlier I myself don't use this feature. Relying on a single "return" buried in `finally` or some `except` smells like bad code to me.  But I'm also not sure how popular this idiom is: I can only recall seeing it once or twice in the wild.

Reply:
I've just pushed an update prohibiting returns in except* altogether.

Reply:
Done in https://github.com/python/exceptiongroups/pull/5/commits/610b4dbed79df4d6b7edc920263f2c5b19670e44